File: LeetCode/two-sum.go
package main

import (
	"fmt"
	"sort"
)

func twoSum(nums []int, target int) (result []int) {
	// class to store value and its original index
	type number struct {
		value int
		index int
	}

	var (
		left, right int
		numsArr     []number
	)

	for i, v := range nums {
		numsArr = append(numsArr, number{
			value: v,
			index: i,
		})
	}

	// sort the elements
	sort.Slice(numsArr, func(i, j int) bool {
		return numsArr[i].value < numsArr[j].value
	})

	// init pointers
	left = 0
	right = len(numsArr) - 1

	for left < right {
		if numsArr[left].value+numsArr[right].value == target {
			result = append(result, numsArr[left].index, numsArr[right].index)
			return
		} else if numsArr[left].value+numsArr[right].value < target {
			left++
		} else { // numsArr[left].value+numsArr[right].value > target
			right--
		}
	}
	return
}

func main() {
	nums := []int{11, 15, 2, 7}
	target := 9
	result := twoSum(nums, target)
	fmt.Println(result)
}

File: LeetCode/two-sum/two-sum.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    struct TwoSum
    {
        int value;
        int index;

        bool operator() (TwoSum a, TwoSum b){
            return a.value < b.value;
        }
    }T;
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<TwoSum> twoSum_;
        vector<int> result;
                
        for(int i=0; i<nums.size(); i++){
            TwoSum temp;
            temp.index = i;
            temp.value = nums[i];
            twoSum_.push_back(temp);
        }

        sort(twoSum_.begin(), twoSum_.end(), T);

        int left = 0;
        int right = twoSum_.size() - 1;

        while (left < right){
            if (twoSum_[left].value + twoSum_[right].value == target){
                result.push_back(twoSum_[left].index);
                result.push_back(twoSum_[right].index);
                break;
            }
            else if (twoSum_[left].value + twoSum_[right].value < target){
                left++;
            }
            else{
                right--;
            }
        }
        return result;
    }
};